sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/crypto/SValueHash.scala

```diff
@@ -0,0 +1,255 @@
+import com.digitalasset.daml.lf.value.Value
+
+import java.nio.ByteBuffer
+import scala.collection.immutable.{ArraySeq, TreeMap}
+
+/** Methods for hashing SValues modulo trailing Nones and package IDs. */
+object SValueHash {
+
+  // --- public interface ---
+
+  /** Hashes a contract key modulo trailing Nones. Throws [[HashingError]] if [key] contains non-serializable values. */
+  @throws[HashingError]
+  def assertHashContractKey(
+      packageName: Ref.PackageName,
+      templateName: Ref.QualifiedName,
+      key: SValue,
+  ): Hash = {
+    builder(Purpose.ContractKey, noCid2String, HashTracer.NoOp)
+      .addString(packageName)
+      .addQualifiedName(templateName)
+      .addSValue(key)
+      .build
+  }
+
+  /** Hashes a contract key modulo trailing Nones. Returns a Left if [key] contains non-serializable values. */
+  def hashContractKey(
+      packageName: Ref.PackageName,
+      templateName: Ref.QualifiedName,
+      key: SValue,
+  ): Either[HashingError, Hash] =
+    Hash.handleError(assertHashContractKey(packageName, templateName, key))
+
+  /** Hashes a contract modulo trailing Nones. Throws [[HashingError]] if [arg] contains non-serializable values. */
+  @throws[HashingError]
+  def assertHashContractInstance(
+      packageName: Ref.PackageName,
+      templateName: Ref.QualifiedName,
+      arg: SValue,
+  ): Hash = {
+    builder(Purpose.ThinContractInstance, aCid2Bytes, HashTracer.NoOp)
+      .addString(packageName)
+      .addQualifiedName(templateName)
+      .addSValue(arg)
+      .build
+  }
+
+  /** Hashes a contract modulo trailing Nones. Returns a Left if [arg] contains non-serializable values. */
+  def hashContractInstance(
+      packageName: Ref.PackageName,
+      templateName: Ref.QualifiedName,
+      arg: SValue,
+  ): Either[HashingError, Hash] =
+    Hash.handleError(assertHashContractInstance(packageName, templateName, arg))
+
+  // --- implementation ---
+
+  private def builder(
+      purpose: Purpose,
+      cid2Bytes: Value.ContractId => Bytes,
+      hashTracer: HashTracer,
+  ): SValueHashBuilder =
+    new SValueHashBuilder(purpose, cid2Bytes, hashTracer).addVersion
+
+  private[crypto] final class SValueHashBuilder(
+      purpose: Purpose,
+      cid2Bytes: Value.ContractId => Bytes,
+      hashTracer: HashTracer,
+  ) extends Builder(bigIntNumericToBytes) {
+
+    private val INT64_TAG: Byte = 0
+    private val NUMERIC_TAG: Byte = 1
+    private val TEXT_TAG: Byte = 2
+    private val TIMESTAMP_TAG: Byte = 3
+    private val PARTY_TAG: Byte = 4
+    private val BOOL_TAG: Byte = 5
+    private val UNIT_TAG: Byte = 6
+    private val DATE_TAG: Byte = 7
+    private val RECORD_TAG: Byte = 8
+    private val VARIANT_TAG: Byte = 9
+    private val ENUM_TAG: Byte = 10
+    private val LIST_TAG: Byte = 11
+    private val OPTIONAL_TAG: Byte = 12
+    private val TEXT_MAP_TAG: Byte = 13
+    private val GEN_MAP_TAG: Byte = 14
+    private val CONTRACT_ID_TAG: Byte = 15
```

Comment:
constants should go in singleton objects, either a companion object or the parent object. 
Here you will allocate and initialize those at each time you construct the object.